engines/immortal/kernal.cpp

```diff
@@ -0,0 +1,1107 @@
+	/* Technically the way this works in the source is that it loads the file
+	 * to a destination address, and then checks the start of that address, and
+	 * if it needs to uncompress, it gives that same address to the uncompress
+	 * routine, overwriting the file in it's place. This is of course slightly
+	 * different here, for simplicity we are not overwriting the original file
+	 * pointer, instead just returning either a compressed or uncompressed
+	 * file pointer.
+	 */
+
+	Common::File f;
+	if (!f.open(fileName)) {
+		debug("*surprised pikachu face*");
+		return nullptr;
+	}
+
+	/* This isn't the most efficient way to do this (could just read a 32bit uint and compare),
+	 * but this makes it more obvious what the source was doing. We want to know if the 4 bytes
+	 * at file[8] are 'C' 'M' 'P' '0', so this grabs just the ascii bits of those 4 bytes,
+	 * allowing us to directly compare it with 'CMP0'.
+	 */
+	char compSig[] = "CMP0";
+		char sig[] = "0000";
+
+	f.seek(8);
+
+	for (int i = 0; i < 4; i++) {
+		sig[i] = f.readByte() & kMaskASCII;
+	}
+
+	if (strcmp(sig, compSig) == 0) {
+		debug("compressed");
+		
+		/* The size of the compressed data is stored in the header, but doesn't
+		 * account for the FORM part?? Also, **technically** this is a uint32LE,
+		 * but the engine itself actually /doesn't/ use it like that. It only
+		 * decrements the first word (although it compares against the second half,
+		 * as if it is expecting that to be zero? It's a little bizarre).
+		 */
+		f.seek(6);
+		int len = f.readUint16LE() - 4;
+
+		// Compressed files have a 12 byte header before the data
+		f.seek(12);
+		return unCompress(&f, len);
+	}
+	// Gotta remember we just moved the cursor around a bunch, need to reset it to read the file
+	f.seek(SEEK_SET);
+
+	byte *out = (byte *)malloc(f.size());
+	f.read(out, f.size());
+	return new Common::MemoryReadStream(out, f.size(), DisposeAfterUse::YES);
+}
+
+
+/*
+ *
+ * -----                   -----
+ * ----- Palette Functions -----
+ * -----                   -----
+ *
+ */
+
+/* Palettes on the Apple IIGS:
+ * In High-res mode you have 2 options: 320x200 @ 4bpp or 320x640 @ 2bpp.
+ * The Immortal uses the former, giving us 16 colours to use
+ * for any given pixel on the screen (ignoring per scanline palettes because
+ * The Immortal does not use them). This 16 colour palette is made of 2 byte
+ * words containing the RGB components in the form 0RGB.
+ *
+ * The equivalent palette for ScummVM is a byte stream of up to 256
+ * colours composed of 3 bytes each, ending with a transparency byte.
+ *
+ * Because each colour in the game palette is only a single nyble (4 bits),
+ * we also need to multiply the nyble up to the size of a byte (* 16, or << 4).
+ */
+
+void ImmortalEngine::loadPalette() {
+	// The palettes are stored at a particular location in the disk, this just grabs them
+	Common::File d;
+	d.open("IMMORTAL.dsk");
+	
+	d.seek(kPaletteOffset);
+	d.read(_palDefault, 32);
```

Comment:
This is not portable. Those values are uint16, so you must read them with:

```C
for (int i = 0; i < 16; i++) {
    _paletteDefault[i] = d.readUint16LE();
}
```

and in the same way for all the palettes below.